File: cc/041.cc
// Project Euler Problem 041
//
// What is the largest n-digit pandigital prime that exists?

#include <cstdio>
#include "primes1.h"
#include "mymath.h"
#include <climits>

// 1+...+8 = 36, and 1+...+9 = 45, so all 8- and 9-digit pandigitals
// are divisible by 3.

int check(int n) {
  int d = digits(n);
  if (d&1) return false;  // 0 not allowed
  d >>= 1;
  while (d) {
    if (!(d&1)) return false;  // missing digits not allowed
    d >>= 1;
  }
  return true;
}

int main(int argc, const char* argv[]) {
  Primes1<int> primes;
  int max = 0;
  int p;
  while ((p=primes.next()) < 1e7) {
    if (p > 2143 && check(p)) {  // 2143 given in problem
      max = p;
    }
  }
  printf("%d\n", max);
}

File: go/problems/041.go
// Project Euler Problem 041
//
// What is the largest n-digit pandigital prime that exists?

package problems

import (
	"fmt"

	"../primes"
)

func pandigitalN(n int) bool {
	mask := 0
	count := uint(0)
	for ; n > 0; n /= 10 {
		mask |= (1 << uint(n%10))
		count++
	}
	return mask == (1<<(count+1) - 2)
}

func Problem041() string {
	max := 0
	for n := 0; ; n++ {
		p := primes.Get(n)
		// sum(1..9) == 45, so 1..8 and 1..9 are divisible by 3
		if p > 7654321 {
			break
		}
		if pandigitalN(p) {
			max = p
		}
	}
	return fmt.Sprintf("%d", max)
}

func init() {
	Functions[41] = Problem041
}
